### DISCUSSION TITLE
[REQ] :: pass variable to retry in useMutation for analytics

### DISCUSSION THREAD

**Discussion by chanphiromsok** [TanStack/query#8217] [Ideas]
I can see in `useMutationState` but it is not reliable as a pass prop to `retry` directly I want to get a variable pass from `mutate` or `mutateAsync` I want to make analytics to firebase and it is hard to depend on variables I have to custom `useCallback` to make useMutationState reflect changing the value in `retry`

![image](https://github.com/user-attachments/assets/096255d9-2a88-4c20-9a30-535429860158)

**Comment by TkDodo** [comment]
retry isn’t the right place for this, it’s supposed to be a pure function that determines if a retry should take place. Depending on what kind of analytics you want, you can set-up a global `onError` handler in the `mutationCache`. I think that’s what you want given that your snippet shows you only want to send something once?

**Reply by chanphiromsok** [reply]
I use useMutationState for case retry and it only captures the first value (does not reactive change) so I use useNonReactiveCallback to custom retry function instead I know useNonReactiveCallback is not good it will capture all objects but this is required (somehow this code is work for my current use case the only things I don't like is I have to put extra effort adding custom reactive function ) to capture variables from useMutationState
I see onError never fired in the case where the internet is slow in my test case 
```js
 const navigation = useNavigation();
  const variables = useMutationState({
    filters: { mutationKey: [MutateKeys.....................] },
    select: (mutation) => mutation.state.variables,
  });
  const variable = variables?.[0];
  const retry = useNonReactiveCallback((failureCount: number, error: AxiosError) => {
    const axiosError = error as AxiosError;
    const statusCode = axiosError?.response?.status;
    const orderId = variable?.orderId || 'unknown';
    if (error?.message === 'Network Error') {
      if (failureCount === 0) {
       crashlytics().recordError(error, `..............${orderId}`);
      }
       return true;
    }

     return failureCount <= 20;
  });

  return useMutation({
    mutationKey: [QueryKeys.mutateAcceptOrder],
    //pass order for dispatching redux if not using redux can safely remove order property
    mutationFn: async ({ order: _, ...params }: AcceptOrderMutateParams) => ............,
    retryDelay: // custom retry,
    retry
  });
  ```

**Reply by TkDodo** [reply]
> and it only captures the first value (does not reactive change)

that shouldn’t be the case. can you show a reproduction of that please?
